ENH: improved error message `IndexError: too many indices for array`

LaisaLI opened the pull request:
improved error message when the dimension of the index is larger than the dimension of the array

brandtbucher committed:
TST: Add failing regression test for unsafe casts.

Errors raised when casting some strided arrays are not caught.
See gh-15790.

sgasse committed:
TST: Add unit test for kwarg of np.einsum

Ensure that explicitly stating out=None does not raise an error in
np.einsum, see #15776 and #15256.

eric-wieser committed:
DEP: Make `np.insert` and `np.delete` on 0d arrays with an axis an error

Before this change, the following code worked:

```
>>> some_0d = np.array(1)
>>> np.insert(some_0d, "some nonsense", 10, axis=0)
array(10)
>>> np.insert(some_0d, "some nonsense", 42, axis="some nonsense")
array(42)
```

Now these raise AxisError and TypeError, respectively.

`delete` is exactly the same.

eric-wieser committed:
DEP: Forbid passing non-integral index arrays to `insert` and `delete`

This expires a deprecation warning from back in 1.9.

eric-wieser committed:
DEP: Make np.delete on out-of-bounds indices an error

Note that this only affects lists of indices.

```python
>>> a = np.arange(3)
````
Before:
```python
>>> np.delete(a, 100)
IndexError
>>> np.delete(a, [100])
DeprecationWarning
array([0, 1, 2])

>>> np.delete(a, -1)
array([0, 1])
>>> np.delete(a, [-1])
FutureWarning
array([0, 1, 2])
```

After:
```python
>>> np.delete(a, 100)
IndexError
>>> np.delete(a, [100])
IndexError

>>> np.delete(a, -1)
array([0, 1])
>>> np.delete(a, [-1])
array([0, 1])
```

mattip committed:
TST: switch PyPy job with CPython [skip travis] (#15809)

seberg committed:
Merge pull request #15805 from eric-wieser/expired-insert-delete-TypeError

DEP: Forbid passing non-integral index arrays to `insert` and `delete`

seberg committed:
Merge pull request #15804 from eric-wieser/expire-delete-out-of-bounds

DEP: Make np.delete on out-of-bounds indices an error

eric-wieser committed:
TST: Remove code that is not supposed to warn out of warning assertion

seberg committed:
Merge pull request #15812 from eric-wieser/expire-delete-out-of-bounds

TST: Remove code that is not supposed to warn out of warning assertion

brandtbucher committed:
MAINT: Break out duplicated logic for strided casts.

brandtbucher committed:
BUG: Break on errors when performing strided casts.

Closed gh-15790.

Commit:
improved error message when the dimension of index is larger than the dimension of array, fixed issue #15321
Author:  Yilin Li <yilin.lasia@gmail.com>

LaisaLI commented:
I found another issue:
```
y = np.arange(35).reshape(5,7)
bool_index = np.array([[True, False, False, True, False]])
y[bool_index, 1]
```
This code also has the same error message but the indexing object is also 2 dimensions.

When I change the code to 
```
# reshape the bool_index
y[bool_index.reshape((5,)), 1]
# or change the type of bool_index
y[bool_index.astype(int), 1]
```
Both above ways are okay.
I think this is a different error.
This one should be some error like `The shape of the boolean index doesn't match the shape of the array `

cournape commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -727,8 +727,14 @@ prepare_index(PyArrayObject *self, PyObject *index,
         }
     }
     else if (used_ndim > PyArray_NDIM(self)) {
+	char error_message[50];
```
nitpick: make sure the indentation is consistent. IIRC, we use 4 space for indentation in the C code.

cournape commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -727,8 +727,14 @@ prepare_index(PyArrayObject *self, PyObject *index,
     }
     else if (used_ndim > PyArray_NDIM(self)) {
+	char error_message[50];
+        PyOS_snprintf(error_message, sizeof(error_message),
```
I think it is better to use `PyErr_Format`:

1. You don't have to create a buffer `error_message` that may not be big enough and causes a failure in `PyOS_snprintf`
2. The code is actually shorter.

Most of the code uses the `PyErr_Format` idiom as well.

cournape commented on numpy/core/tests/test_indexerrors.py:
```diff
@@ -110,6 +111,13 @@ def assign(obj, ind, val):
 
+    def test_mapping_error_message(self):
+        a = np.zeros((3, 5))
+        index =(2, 4, 0)
```
nitpick: missing space after `=`

cournape commented on numpy/core/tests/test_indexerrors.py:
```diff
@@ -110,6 +111,13 @@ def assign(obj, ind, val):
+        a = np.zeros((3, 5))
+        index =(2, 4, 0)
+        with pytest.raises(IndexError,
+                match=f"array has {a.ndim}-dimension, but {len(index)} were indexed"):
```
same here: for styling, your code editor likely has some options to automatically check for PEP8 in real time. If you don't know, let me know, and I can help you setting it up.

cournape commented on numpy/core/tests/test_indexerrors.py:
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 from numpy.testing import assert_raises
+import pytest
```
If you look at existing tests, you will notice they use `assert_raises`, and do not import `pytest1 directly.

cournape reviewed (changes requested):
The basic looks good, but some style nitpicks and changes.

In addition, it will be good for the PR to refer to the github issue you're gonna fix (e.g. add something like `Fixes #<case number>`).

cournape commented:
>Both above ways are okay.
> I think this is a different error.
> This one should be some error like The shape of the boolean index doesn't match the shape of the array

This is a different issue. Basically, numpy has two modes of indexing: the 'normal' one and so-called fancy indexing. Fancy indexing happens as soon as you use an array `x` to index an array `a`, e.g. `a[x]`. So when you do stuff like:

```
y = np.arange(35).reshape(5,7)
bool_index = np.array([[True, False, False, True, False]])
y[bool_index, 1]

# reshape the bool_index
y[bool_index.reshape((5,)), 1]
# or change the type of bool_index
```

You are using fancy indexing. When using fancy indexing, NumPy will try to "broadcast" the indices arrays to match the array being indexed. You can see here https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html, and also there for advanced discussions: https://numpy.org/neps/nep-0021-advanced-indexing.html. Fancy indexing is quite tricky, lots of weird corner cases. It will melt your mind :)

eric-wieser committed:
Merge pull request #15802 from eric-wieser/simplify-insert

DEP: Make `np.insert` and `np.delete` on 0d arrays with an axis an error

eric-wieser committed:
Merge pull request #15800 from brandtbucher/unsafe-cast

BUG: Add error-checking versions of strided casts.

rth committed:
pathlib is in the standard library

rth committed:
md5 module is deprecated since python 2.5

WarrenWeckesser committed:
BUG: Add basic __format__ for masked element to fix incorrect print (#15794)

Fix and test conversion of masked element to string in fstring / string interpolation. See PEP 498 for string interpolation (aka fstrings) in Python 3.6 and above.

Fixes #15409.

Original PR gh-15410 by: Stefan Codrescu <ssmmcc1@gmail.com>

Co-authored-by: Stefan Codrescu <ssmmcc1@gmail.com>
Co-authored-by: Stefan <5tefan@users.noreply.github.com>
Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

eric-wieser committed:
Merge pull request #15830 from rth/import-fixes-mypy

MAINT: pathlib and hashlib are in stdlib in Python 3.5+

brandtbucher committed:
BUG: Break on all errors when performing strided casts. (gh-15827)

Errors raised when casting some strided arrays are not caught.
See gh-15790.

seberg committed:
Merge pull request #15797 from sgasse/add_einsum_test

TST: Add unit test for out=None of np.einsum

eric-wieser committed:
DEP: Do not cast boolean indices to integers in np.delete (#15815)

This expires a deprecation from 1.8.

The corresponding deprecation in `np.insert` has less clear semantics, so has been left to a future patch.

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>
Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

cournape commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -727,8 +727,11 @@ prepare_index(PyArrayObject *self, PyObject *index,
-        PyErr_SetString(PyExc_IndexError,
-                        "too many indices for array");
+        PyErr_Format(PyExc_IndexError,
+		     "array has %"NPY_INTP_FMT
```
Still indentation error: which editor are you using ?

LaisaLI replied:
vim...

LaisaLI replied:
In local, there's no indent error. That's really weird...

![image](https://user-images.githubusercontent.com/4524261/77511910-560c3d00-6eb5-11ea-8911-c36521eb9291.png)

cournape commented on numpy/core/tests/test_indexerrors.py:
```diff
@@ -110,6 +111,16 @@ def assign(obj, ind, val):
+    def test_mapping_error_message(self):
+        a = np.zeros((3, 5))
+        index = (1, 2, 3, 4, 5)
+        try:
```
You should use `assert_raises`

LaisaLI replied:
As far as I know, `assert_raise` can not get the `error message`.

cournape commented on numpy/core/tests/test_indexerrors.py:
```diff
@@ -1,5 +1,8 @@
 import numpy as np
-from numpy.testing import assert_raises
+from numpy.testing import (
```
I think you can just do `from numpy.testing import assert_raises, assert_raises_regex`, it should still be in the limit.

cournape reviewed (approved):
Just one last nitpick, but LGTM otherwise.

LaisaLI committed:
ENH: changed the error message to make it more clear

LaisaLI committed:
Changed the error message to make it more clear

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

LaisaLI committed:
BUG: changed the test error message to match the error message in mapping.c
